keps/sig-storage/3476-volume-group-snapshot/README.md

```diff
@@ -0,0 +1,1660 @@
+
+It is true that we can quiesce the application first, take an individual snapshot from each volume that is part of the application one after the other, and then unquiesce the application after all the individual snapshots are taken. This way we will get application consistent snapshots. However, application quiesce is time consuming. Sometimes it may not be possible to quiesce an application. Taking individual snapshots one after another may also take longer time compared to taking a consistent group snapshot. Some users may not want to do application quiesce very frequently for these reasons. For example, a user may want to run weekly backups with application quiesce and nightly backups without application quiesce but with consistency group support which provides crash consistency across all volumes in group.
+
+There is also another KEP (https://github.com/kubernetes/enhancements/pull/1051) that introduces APIs to do application snapshot, backup, and restore, but that KEP has a broader scope. In other words, volume group snapshot proposed in this KEP can be used by the application snapshot proposed in the other KEP.
+
+### Goals
+
+* Provide an API to take a snapshot of multiple volumes together.
+
+### Non Goals
+
+* Provide an API to manage multiple volumes together in a group.
+* Provide a generic and extensible group API that may be used to support other features in the future.
+* Provide a VolumeGroup API that supports group replication.
+* Provide a design to facilitate volume placement using the group API.
+
+## Proposal for VolumeGroupSnapshot
+
+This proposal introduces new CRDs VolumeGroupSnapshot, VolumeGroupSnapshotContent, and VolumeGroupSnapshotClass.
+
+### Create VolumeGroupSnapshot
+
+A VolumeGroupSnapshot can be created from multiple PVCs with a label on the PVCs specified by the labelSelector in the VolumeGroupSnapshot if the CSI driver supports the CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT capability.
+
+#### Dynamic provisioning
+
+* Admin creates a VolumeGroupSnapshotClass.
+* User creates a VolumeGroupSnapshot with label selector that matches the label applied to all PVCs to be snapshotted together.
+* This will trigger the VolumeGroupSnapshot controller to create a VolumeGroupSnapshotContent API object, and also call the CreateVolumeGroupSnapshot CSI function. It will also create multiple VolumeSnapshot API objects with volumeGroupSnapshotName in the status and the corresponding VolumeSnapshotContents with the snapshot handle. The VolumeSnapshot and VolumeSnapshotContent will point to each other before these objects are created in the API server to avoid triggering the VolumeSnapshot controller to create new individual objects. The CSI snapshotter sidecar will not call CSI driver in this case. If needed, GetVolumeGroupSnapshot CSI function will be called to retrieve individual snapshot statuses until all snapshots are ready to use.
```

Comment:
Regarding automatically creating VolumeSnapshot objects, how can we deal with ownership and potential conflicts with user-created VolumeSnapshot objects?

Reply:
These VolumeSnapshots are created in the same namespace as VolumeGroupSnapshot.  The controller will bind them before the objects are created.  VolumeSnapshotContents will be created in API server first and they are already pointing to VolumeSnapshot names.  VolumeSnapshots will be created in API server next and pointing back to the contents.  So there should not be conflicts.

Reply:
How will the names be generated? Could there be a name conflict?

Reply:
Updated to create pre-provisioned VolumeSnapshots.